Doc/man/editwin.man
.TH EDITWIN 3
.SH NAME
Editwin \- editing windows package for STDWIN
.SH SYNOPSIS
.nf
.ft C
#include "stdwin.h"
#include "editwin.h"

typedef struct editwin {
	WINDOW *win;
	TEXTEDIT *tp;
	char *filename;
	char saved;
} EDITWIN;

EDITWIN *ewcreate(char *filename);
EDITWIN *ewnew();
EDITWIN *ewopen();

bool ewsave(EDITWIN *ew);
bool ewsaveas(EDITWIN *ew);
bool ewsavecopy(EDITWIN *ew);
bool ewsaveall();
bool ewrevert(EDITWIN *ew);
bool ewclose(EDITWIN *ew);
bool ewcloseall();

bool ewwritefile(EDITWIN *ew, char *filename);
bool ewreadfile(EDITWIN *ew, char *filename);

bool ewevent(EDITWIN *ew, EVENT *e, int *closed_return);
void ewreplace(EDITWIN *ew, char *str);
void ewundo(EDITWIN *ew); /* Not implemented */
void ewcopy(EDITWIN *ew);
void ewpaste(EDITWIN *ew);

EDITWIN *ewfind(WINDOW *win);
int ewcount();
.ft 1
.fi
.SH DESCRIPTION
.I Editwin
is a package built on top of the
.I textedit
package, to ease the construction of views on text files etc.
Many calls exist to make it extremely simple to respond to standard
menus of
.B File
operations (New, Open..., Save, etc.) and
.B Edit
operations (Cut, Copy, Paste).
.PP
Below are descriptions of the individual functions.
Note that when a reference is made to a window's contents, the entire
contents of the edit buffer belonging to the window is meant, not just
the part of it visible in the window.
.IP ewcreate
If
.I filename
is a nil pointer, this call creates an ``Untitled'' edit window.
Otherwise, the specified file must exist, and an edit window showing
its contents is created.
In both cases, the window's `saved' flag is set initially.
A pointer to a newly allocated EDITWIN struct is returned.
If something went wrong (e.g., insufficient memory, or an unreadable
file was specified), a nil pointer is returned.
.IP ewnew
This function can be called in response to the selection of the
.B New
menu item.
It is equivalent to a call to
.I ewcreate((char*)0).
.IP ewopen
Call this function for the
.B Open...
menu item.
It asks the user for an existing file name (using
.IR waskfile ),
and then calls
.I ewcreate
with that file name as parameter.
It returns nil if the dialog was cancelled or
.I ewcreate
returns nil.
.IP ewsave
Call this function for the
.B Save
menu item.
If the window's contents were modified since they were last read or
saved, the function attempts to save the window to its file.
If the window was still ``Untitled'', the user is first asked to specify
a file name.
The function returns true (nonzero) if the contents were actually saved,
or didn't need saving.
.IP ewsaveas
Call this function for the
.B Save As...
menu item.
It asks the user for a new file name to save the window's contents, and
if the saving succeeds, sets this to be the file name to which future
save operations are directed.
.IP ewsavecopy
Call this function for the
.B Save a Copy...
menu item.
Like
.IR ewsaveas ,
this function asks file a new file name and saves the window's contents
to that file; but it does not change the file name used for future save
operations.
.IP ewsaveall
Calls
.I ewsave
for all windows.
If any call returns false (zero),
.I ewsaveall
skips further calls and returns false.
.IP ewrevert
Call this function for the
.B Revert...
menu item.
It attempts to undo any changes since the window was last read or
saved, by re-reading the corresponding file.
If this is at all possible, the user is asked to confirm the operation
first (since it may destroy valuable changes).
The function returns true if the file was actually read back, or if the
window was unchanged with respect to the file.
.IP ewclose
Closes the window.
If the window was changed since it was last read or saved, the user is
first asked whether it should be saved, and if the answer is Yes,
.I ewsave
is called.
Cancelling the dialog will prevent closing the window.
Returns true if the window was actually closed.
.IP ewcloseall
Calls
.I ewclose
for all windows.
If any call returns false,
.I ewcloseall
skips further calls and returns false.
.IP ewwritefile
Writes the contents of the edit window to the specified file.
Returns true if the operation succeeded.
This does
.I not
set the `saved' flag for the window (because it is used internally be
.IR ewsavecopy ).
.IP ewreadfile
Reads the contents of the given file into the edit window, discarding
its previous contents.
Returns true if the operation succeeded.
This
.I does
set the `saved' flag for the window.
.IP ewevent
Call this function in response to
.I any
event returned by
.I wgetevent.
If the event is a non-menu event applicable to the specified window, it
is handled and the function returns true;
otherwise nothing is done and the function returns false.
(Menu events cannot be handled this way because the editwin package
doesn't create its own menus, and thus cannot know the menu IDs or the
numbers of the menu items.)
If the first parameter is a nil pointer, the event is checked against
any edit window; otherwise, only events applying to the given window are
handled.
The third parameter must be a pointer to an integer variable, which is
cleared normally when an event was handled, but set to true when a
window was closed as a consequence of the event (it is unchanged when
the event was nbot handled at all).
In the latter case, the caller should check whether any windows are
still open (see
.IR ewcount ),
and if this is not the case, it should either exit or open a new window.
This function clears a window's `saved' flag whenever its contents are
modified by the event's handling.
.IP ewreplace
Replaces the current text selection in the window by the given
(null-terminated) string.
This will insert text if the text selection was an insert point.
.IP ewundo
Reserved for future extension of the package with an Undo facility.
.IP ewcopy
Call this function for the
.B Copy
menu item.
It retrieves the contents of the window's text selection, if non-null,
and copies it in into the clipboard through a call to
.I wsetclip.
It beeps if the text selection is empty.
(To implement the
.B Cut
menu item, call
.I ewcopy
followed by
.IR "ewreplace(ew, ``'')" .)
.IP ewpaste
Call this function for the
.B Paste
menu item.
It retrieves the contents of the clipboard, if non-null,
and pastes it into the window through a call to
.I ewreplace.
It beeps (and does not change the window's contents) if the clipboard is
empty.
.IP ewfind
Returns a pointer to the EDITWIN struct containing the given WINDOW
pointer; returns nil if none exists.
.IP ewcount
Returns the number of open edit windows.
.SH EXAMPLE
The following program is a trivial but almost usable single-file text
editor.
Usage is ``program [file]''.
.nf
.ft C

#include "stdwin.h"
#include "editwin.h"

main(argc, argv) int argc; char **argv; {
	EDITWIN *ew;
	winitnew(&argc, &argv);
	if (argc <= 1) ewnew();
	else           ewcreate(argv[1]);
	for (;;) {
		EVENT e;
		int closed;
		wgetevent(&e);
		if (ewevent(ew, &e, &b) && closed) break;
	}
	wdone();
	exit(0);
}
.ft 1
.fi
.SH HINTS
The members of the EDITWIN data structure are explicitly intended to be
accessible to the caller.
Functionality which is not provided directly but which is available for
plain windows or for textedit data structures can be implemented by
applying it to the
.I win
or
.I tp
members.
Note that the
.I filename
member, when non-nil, points to memory allocated with
.IR malloc (3).
.PP
Changes to the window's contents should preferably be made with
.I ewreplace,
since it manages the `saved' flag.
.PP
To control the text attributes used in an EDITWIN window, you can set
them globally before creating the window.
.SH DIAGNOSTICS
.I Ewcreate, ewnew
and
.I ewopen
return nil when the user cancelled the operation or when they could not
get all the necessary memory.
The save, revert, close and read/write family of functions return FALSE
if the operation was canceled by the user or if the file I/O failed.
.I Ewevent
returns TRUE when it has processed the event.
.PP
.I Ewcreate
and
.I ewopen
warn the user if a file is larger than about 30K; the textedit package
was not designed to operate on big files, and may be intolerably slow.
.SH SEE ALSO
STDWIN documentation
.br
textedit(3)
.SH AUTHOR
Guido van Rossum
.SH BUGS
.I Editwin
inherits some bugs from the
.I textedit
package.
.br
The package doesn't detect the situation where the user opens the same
file twice, edits both copies, and saves them, thus losing the
changes to the copy saved first.
.br
If the height of the document gets over 32K scan lines, you are
in trouble: most window systems limit coordinates to short integers.
.br
Missing functionality:
a way to specify an alternate title for an untitled window;
a way to create a non-file window, which isn't saved when closed.
(These should be easy to add, given the simplicity of the source.)


Packs/textedit/editwin.c
/* Edit Windows */

/* The data structure exported by this module is intended to
   be accessible to the caller;
   the routines provided here don't duplicate operations that
   can be done by calling text-edit or window routines directly.
   Exception: ewreplace is like tereplace, but also clears
   the 'saved' flag and sets the document size */

#include "stdwin.h"
#include "tools.h"
#include "editwin.h"

extern long ftell _ARGS((FILE *));

#define MAXFN 256 /* File name length */

static int ewnum;
static EDITWIN **ewlist;

EDITWIN *
ewfind(win)
	WINDOW *win;
{
	int i;

	if (win == NULL)
		return NULL;

	i= wgettag(win);
	
	if (i >= 0 && i < ewnum &&
			ewlist[i] != NULL && ewlist[i]->win == win)
		return ewlist[i];
	else
		return NULL;
}

int
ewcount()
{
	int count= 0;
	int i;
	for (i= 0; i < ewnum; ++i) {
		if (ewlist[i] != NULL)
			++count;
	}
	return count;
}

static void
ewdrawproc(win, left, top, right, bottom)
	WINDOW *win;
{
	EDITWIN *ew= ewfind(win);
	if (ew != NULL)
		tedrawnew(ew->tp, left, top, right, bottom);
}

EDITWIN *
ewcreate(filename)
	char *filename;
{
	EDITWIN *ew= ALLOC(EDITWIN);
	int width, height;
	int hbar, vbar;
	int i;
	
	if (ew == NULL)
		return NULL;
	wgetdefscrollbars(&hbar, &vbar);
	wsetdefscrollbars(0, 1);
	ew->win= wopen(filename==NULL ? "Untitled" : filename, ewdrawproc);
	wsetdefscrollbars(hbar, vbar);
	if (ew->win == NULL) {
		FREE(ew);
		return NULL;
	}
	wsetwincursor(ew->win, wfetchcursor("ibeam"));
	wgetwinsize(ew->win, &width, &height);
	ew->tp= tecreate(ew->win, 0, 0, width, height);
	if (ew->tp == NULL) {
		wclose(ew->win);
		FREE(ew);
		return NULL;
	}
	if (filename != NULL) {
		if (!ewreadfile(ew, filename)) {
			tefree(ew->tp);
			wclose(ew->win);
			FREE(ew);
			return NULL;
		}
	}
	ew->filename= strdup(filename);
	ew->saved= TRUE;
	for (i= 0; i < ewnum; ++i) {
		if (ewlist[i] == NULL)
			break;
	}
	wsettag(ew->win, i);
	if (i >= ewnum) {
		L_APPEND(ewnum, ewlist, EDITWIN*, NULL);
	}
	ewlist[i]= ew;
	return ew;
}

EDITWIN *
ewnew()
{
	return ewcreate((char*)NULL);
}

EDITWIN *
ewopen()
{
	char filename[MAXFN];
	
	filename[0]= EOS;
	if (waskfile("Open file:", filename, sizeof filename, FALSE))
		return ewcreate(filename);
	else
		return NULL;
}

bool
ewclose(ew)
	EDITWIN *ew;
{
	int i;
	
	if (!ew->saved) {
		char buf[MAXFN+25];
		sprintf(buf, "Save changes to \"%s\" before closing?",
			ew->filename==NULL ? "Untitled" : ew->filename);
		switch (waskync(buf, 1)) {
		case -1:
			return FALSE;
		case 0:
			break;
		case 1:
			if (!ewsave(ew))
				return FALSE;
			break;
		}
	}
	i= wgettag(ew->win);
	if (i >= 0 && i < ewnum && ewlist[i] == ew)
		ewlist[i]= NULL;
	tefree(ew->tp);
	wclose(ew->win);
	FREE(ew->filename);
	FREE(ew);
	return TRUE;
}

bool
ewsave(ew)
	EDITWIN *ew;
{
	if (ew->saved)
		return TRUE;
	if (!ew->filename)
		return ewsaveas(ew);
	return ew->saved= ewwritefile(ew, ew->filename);
}

bool
ewsaveas(ew)
	EDITWIN *ew;
{
	return ewsaveprompt(ew, "Save as:", TRUE);
}

bool
ewsavecopy(ew)
	EDITWIN *ew;
{
	return ewsaveprompt(ew, "Save a copy as:", FALSE);
}

bool
ewsaveprompt(ew, prompt, changefile)
	EDITWIN *ew;
	char *prompt;
	bool changefile;
{
	char filename[MAXFN];
	
	filename[0]= EOS;
	if (ew->filename != NULL) {
		strncpy(filename, ew->filename, MAXFN);
		filename[MAXFN-1]= EOS;
	}
	if (!waskfile(prompt, filename, sizeof filename, TRUE))
		return FALSE;
	if (!ewwritefile(ew, filename))
		return FALSE;
	if (changefile) {
		FREE(ew->filename);
		ew->filename= strdup(filename);
		wsettitle(ew->win, filename);
		ew->saved= TRUE;
	}
	return TRUE;
}

bool
ewrevert(ew)
	EDITWIN *ew;
{
	char buf[MAXFN+50];
	
	if (ew->saved)
		return TRUE;
	if (ew->filename == NULL) {
		wmessage("Not saved yet");
		return FALSE;
	}
	sprintf(buf, "Discard changes since last save to \"%s\" ?",
		ew->filename);
	if (waskync(buf, 1) < 1)
		return FALSE;
	return ewreadfile(ew, ew->filename);
}

bool
ewreadfile(ew, filename)
	EDITWIN *ew;
	char *filename;
{
	FILE *fp= fopen(filename, "r");
	long size;
	char *buf;
	
	if (fp == NULL) {
		wperror(filename);
		return FALSE;
	}
	if (fseek(fp, 0L, 2) == -1) {
		wperror (filename);
		fclose (fp);
		return FALSE;
	}
	size= ftell(fp);
	if (size >= 1L<<15) {
		char mbuf[MAXFN + 50];
		sprintf(mbuf, "File \"%s\" is big (%d K).  Still edit?",
			filename, (int) ((size+1023)/1024));
		if (waskync(mbuf, 1) < 1) {
			fclose(fp);
			return FALSE;
		}
	}
	buf= malloc((unsigned)size);
	if (buf == NULL) {
		wmessage("Can't get memory for buffer");
		fclose(fp);
		return FALSE;
	}
	if (fseek(fp, 0L, 0) == -1) {
		wperror (filename);
		fclose (fp);
		return FALSE;
	}
	if ((size = fread(buf, 1, (int) size, fp)) == -1) {
		wmessage("Read error");
		fclose(fp);
		return FALSE;
	}
	fclose(fp);
	tesetbuf(ew->tp, buf, (int) size); /* Gives the buffer away! */
	ew->saved= TRUE;
	ewsetdimensions(ew);
	return TRUE;
}

void
ewsetdimensions(ew)
	EDITWIN *ew;
{
	wsetdocsize(ew->win, 0, tegetbottom(ew->tp));
}

bool
ewwritefile(ew, filename)
	EDITWIN *ew;
	char *filename;
{
	FILE *fp= fopen(filename, "w");
	char *buf;
	int len, nwritten;
	
	if (fp == NULL) {
		wperror(filename);
		return FALSE;
	}
	buf= tegettext(ew->tp);
	len= strlen(buf);
	nwritten= fwrite(buf, 1, len, fp);
	fclose(fp);
	if (nwritten != len) {
		wmessage("Write error");
		return FALSE;
	}
	/* Can't set saved to TRUE, because of 'save a copy' */
	return TRUE;
}

bool
ewevent(ew, e, closed_return)
	EDITWIN *ew;
	EVENT *e;
	bool *closed_return;
{
	bool closed= FALSE;
	bool change= FALSE;
	
	if (ew == NULL) {
		ew= ewfind(e->window);
		if (ew == NULL)
			return FALSE;
	}
	else if (e->window != ew->win)
		return FALSE;
	
	switch (e->type) {
	
	case WE_ACTIVATE:
		break;
		
	case WE_SIZE:
		{
			int width, height;
			wgetwinsize(ew->win, &width, &height);
			temovenew(ew->tp, 0, 0, width, height);
			ewsetdimensions(ew);
		}
		break;
		
	case WE_CLOSE:
		closed= ewclose(ew);
		break;

	case WE_COMMAND:
		if (e->u.command == WC_CLOSE) {
			closed= ewclose(ew);
			break;
		}
		/* Else, fall through */
		if (e->u.command == WC_RETURN ||
			e->u.command == WC_TAB ||
			e->u.command == WC_BACKSPACE)
			change= TRUE;
		goto def; /* Go pass it on to teevent */
	
	case WE_MOUSE_DOWN:
		/* Edit the coordinates slightly so teevent always
		   believes it is in the rect */
		{
			int left= tegetleft(ew->tp);
			int right= tegetright(ew->tp);
			int top= tegettop(ew->tp);
			int bottom= tegetbottom(ew->tp);
			CLIPMIN(e->u.where.h, left);
			CLIPMAX(e->u.where.h, right);
			CLIPMIN(e->u.where.v, top);
			if (e->u.where.v >= bottom) {
				e->u.where.v= bottom;
				e->u.where.h= right;
			}
		}
		/* Fall through */
	default:
	def:
		if (!teevent(ew->tp, e))
			return FALSE;
		if (e->type == WE_CHAR)
			change= TRUE;
		if (change) {
			ew->saved= FALSE;
			ewsetdimensions(ew);
		}
		break;
		
	}
	
	if (closed_return != NULL)
		*closed_return= closed;
	return TRUE;
}

bool
ewsaveall()
{
	int i;
	
	for (i= 0; i < ewnum; ++i) {
		if (ewlist[i] != NULL && !ewsave(ewlist[i]))
			return FALSE;
	}
	return TRUE;
}

bool
ewcloseall()
{
	int i;
	
	for (i= 0; i < ewnum; ++i) {
		if (ewlist[i] != NULL && !ewclose(ewlist[i]))
			return FALSE;
	}
	return TRUE;
}

void
ewreplace(ew, str)
	EDITWIN *ew;
	char *str;
{
	if (*str == EOS && tegetfoc1(ew->tp) == tegetfoc2(ew->tp))
		return;
	tereplace(ew->tp, str);
	ew->saved= FALSE;
	ewsetdimensions(ew);
}

/*ARGSUSED*/
void
ewundo(ew)
	EDITWIN *ew;
{
	wfleep();
}

void
ewcopy(ew)
	EDITWIN *ew;
{
	int f1= tegetfoc1(ew->tp);
	int f2= tegetfoc2(ew->tp);
	char *text;
	if (f1 == f2)
		wfleep();
	else {
		text= tegettext(ew->tp);
		wsetclip(text+f1, f2-f1);
	}
		
}

void
ewpaste(ew)
	EDITWIN *ew;
{
	char *text= wgetclip();
	if (text == NULL)
		wfleep();
	else
		ewreplace(ew, text);
}
